backend/src/pong/pong.service.ts

```diff
@@ -0,0 +1,155 @@
+import { Injectable } from '@nestjs/common';
+import { Ball, GameState, MovementKeys, Paddle, Position } from './Game.types';
+
+@Injectable()
+export class PongService {
+	private readonly paddleWidth = 1;
+	private readonly paddleHeight = 12;
+
+	private readonly ballRadius = 1;
+
+	// ball direction and speed %
+	private dx = -0.4;
+	private dy = 0.4;
+
+	createNewGameState(): GameState {
+		const newGameState: GameState = {
+			playerOnePaddle: {
+				position: {
+					x: 0,
+					y: 50,
+				},
+				width: this.paddleWidth,
+				height: this.paddleHeight,
+			},
+
+			playerTwoPaddle: {
+				position: {
+					x: 100,
+					y: 50,
+				},
+				width: this.paddleWidth,
+				height: this.paddleHeight,
+			},
+
+			ball: {
+				position: {
+					x: 40,
+					y: 65,
+				},
+				radius: this.ballRadius,
+			},
+			canvas: {
+				x: 1200,
+				y: 960,
+			},
+		};
+		return newGameState;
+	}
+
+	doesHitWall(radius: number, pos: number, direction: number): boolean {
+		const fullHeigthOrWidth = 100;
+		return (
+			pos + direction > fullHeigthOrWidth - radius || pos + direction < radius
+		);
+	}
+
+	doesHitPaddle(ball: Ball, rect: Paddle, canvas: Position) {
+		const paddleY = rect.position.y;
+		const paddleHeight = rect.height;
+		const paddleTop = paddleY - paddleHeight / 2;
+		const paddleBottom = paddleY + paddleHeight / 2;
+		const magicOffset = 0.2;
+		const paddleWall = ball.radius - magicOffset + this.paddleWidth / 2;
+
+		// if ball will hit the invisible wall that extends from the paddle
+		// and ball position is within the paddle top and bottom
+		// console.log(paddleWall);
+		if (
+			this.doesHitWall(paddleWall, ball.position.x, this.dx) &&
+			ball.position.y > paddleTop &&
+			ball.position.y < paddleBottom
+		) {
+			return true;
+		}
+
+		return false;
+	}
+
+	// moves ball dy(directioY) / x(directionX)%
+	// if ball hits a wall the directiion (dx /dy) reverses
+	// with game logic implemented, hitting the y axis-wall would be a point
+	moveBall(gameState: GameState) {
+		const radius = gameState.ball.radius;
+		const ballPos: Position = gameState.ball.position;
+		const paddleP1: Paddle = gameState.playerOnePaddle;
+		const paddleP2: Paddle = gameState.playerTwoPaddle;
+
+		//  check collision for paddle one
+		if (this.doesHitPaddle(gameState.ball, paddleP1, gameState.canvas)) {
+			console.log('Collision Paddle one');
+			const paddleCenter = paddleP1.position.y + paddleP1.height / 2;
+			const d = paddleCenter - ballPos.y;
+			//  speed up the ball
+			this.dx = -(this.dx - 0.02);
+			// change bounce angle
+			this.dy -= d * -0.01;
```

Comment:
does the ball increasingly become a little bit faster without limit?
and same-ish question about the angle

i haven't played it, can also be fine since the values are quite low, just curious how it'll play

Reply:
The speed does, the angle doesn't. "d" value for the changing of the y-angle ranges from +-"0.01 - 0.7" but changes for every impact.

As for the speed, I can limit it, but it increases very slowly, so by the time it's quite fast the point has been going for some time already.

Reply:
If the speed does not increase at an unreasonable rate I'm fine with it :P It should probably at some point become too fast so we "force" a point to make the games not infinite